src/control/lib/telemetry/promexp/collector.go

```diff
@@ -43,17 +43,21 @@ type (
 	labelMap map[string]string
 )
 
-func NewEngineSource(parent context.Context, idx uint32, rank uint32) (*EngineSource, error) {
+func NewEngineSource(parent context.Context, idx uint32, rank uint32) (*EngineSource, func(), error) {
 	ctx, err := telemetry.Init(parent, idx)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to init telemetry")
+		return nil, nil, errors.Wrap(err, "failed to init telemetry")
+	}
+
+	cleanupFn := func() {
+		telemetry.Detach(ctx)
 	}
 
 	return &EngineSource{
 		ctx:   ctx,
```

Comment:
The collector is in need of some general refactoring. That having been said, the reason we can't easily change this is because the Collector implements the [prometheus.Collector](https://github.com/prometheus/client_golang/blob/master/prometheus/collector.go#L62) interface which doesn't give us an opportunity to provide a context via dependency injection.

As we discussed before, the primary idea behind that advice is to encourage use of contexts by dependency injection in order to allow for more decoupled code (e.g. if we need a function to run with a timeout we can wrap its context before calling it rather than having to change the function to implement a timeout) that's easier to read/maintain. There isn't anything inherently wrong or dangerous about storing a context in a struct, it's just considered a bad code pattern in most cases.